### DISCUSSION TITLE
Querying Role Assignment Data

### DISCUSSION THREAD

**Discussion by kentlhastings** [pnp/pnpcore#371] [Q&A] [answered]
Hello all,

I would like to query a particular folder and read its RoleAssignments in order to assign the same roles to some other folders and files, but this seems to require a call to QueryOptions as it's a nested property from the parent folder.

What is the proper way to execute this query, if ListItemAllFields doesn't have a QueryOptions member?

var source = await context.Web.GetFolderByServerRelativeUrlAsync(folderPath, x => x.ListItemAllFields.RoleAssignments);

Thank you!

**Comment by jansenbe** [comment, marked as answer]
You can use the QueryProperties LINQ method to realize this:

```csharp
var source = await context.Web.GetFolderByServerRelativeUrlAsync(folder.ServerRelativeUrl, x => x.ItemCount, x => x.Folders, x => x.Files, x => x.ListItemAllFields.QueryProperties(p => p.RoleAssignments));
```

**Reply by kentlhastings** [reply]
I was missing a using statement that allows ListItemAllFields to have the QueryProperties member. Thank you!

**Reply by kentlhastings** [reply]
@jansenbe Hey Bert, I tried this a couple ways and kept getting an "Unresolved tokens" error. {"Unresolved tokens found in API call sharePointSiteURL/Lists(guid'{List.Id}')/Items(589)/RoleAssignments?$top=100"}

Here are the variations I was attempting:

var source = await batchContext.Web.GetFolderByServerRelativeUrlAsync(sourceFolderPath, x => x.ListItemAllFields.QueryProperties(x => x.RoleAssignments));

                var source = await batchContext.Web.GetFolderByServerRelativeUrlAsync(sourceFolderPath, x => x.ListItemAllFields.QueryProperties(x => x.RoleAssignments.QueryProperties(x => x.RoleDefinitions, x => x.PrincipalId)));

                var source = await batchContext.Web.GetFolderByServerRelativeUrlAsync(sourceFolderPath, x => x.UniqueId, x => x.ServerRelativeUrl, x => x.ListItemAllFields.QueryProperties(x => x.RoleAssignments.QueryProperties(x => x.RoleDefinitions, x => x.PrincipalId)));

**Reply by jansenbe** [reply]
So this does not work?

```csharp
var source = await batchContext.Web.GetFolderByServerRelativeUrlAsync(sourceFolderPath, x => x.ListItemAllFields.QueryProperties(x => x.RoleAssignments));
```

**Reply by kentlhastings** [reply]
Correct, I noticed with some other call I was trying to make yesterday that adding more parameters stopped the "unresolved tokens" error, but no luck in this case. I've been assuming that "unresolved tokens" means that we're trying to query incomplete data because properties are fetched piecemeal?

**Reply by kentlhastings** [reply]
@jansenbe Hi Bert, I was able to go deeper into the issue. The initial query (context.Web.GetFolderByServerRelativeUrlAsync(sourceFolderPath, x => x.ListItemAllFields.QueryProperties(x => x.RoleAssignments)) returns successfully, but we are failing immediately after by using the resulting IFolder.ListItemAllFields.RoleAssignments in a ForEach. When the GetEnumerator is called, the resulting query returns "unresolved tokens" error. I hope this helps!!

**Reply by jansenbe** [reply]
You need to add AsRequested() on the collection before querying it as otherwise we try to do a new server query

**Reply by kentlhastings** [reply]
Thank you!

**Reply by GergiH** [reply]
Regarding the `Unresolved tokens found in API call` issue, I'm still bumping into that, for both `GetRoleDefinitionsAsync()` and `RemoveRoleDefinitionsAsync()`. It actually happens at `.../Lists(guid'{List.Id}')/Items...` in the request url, all the rest of the variable params are there.
What might I be missing?

I'm getting the folder with the relative url, and tried loading the following on the `folder.ListItemAllFields`, but none helped:
```
x => x.ParentList
x => x.ParentList.Id
x => x.ParentList.QueryProperties(y => y.Id)
```

I've also tried with the above mentioned way to load the properties, and with `.LoadAsync()`, no success.

Adding works though, I guess it doesn't need to load the existing permissions on the folder.
